### ISSUE TITLE
[TODO]

### ISSUE THREAD

**Issue by alessandro-maccario** [alessandro-maccario/DeepLearning-FoodRecognition#1]
Check:
- https://www.quora.com/How-can-I-improve-my-validation-accuracy-in-my-CNN-model
- https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html
- https://towardsdatascience.com/integration-of-dimension-reduction-methods-and-neural-network-for-image-classification-96281963fe24
- https://towardsdatascience.com/using-pca-to-reduce-number-of-parameters-in-a-neural-network-by-30x-times-fcc737159282
-  https://wandb.ai/wandb_fc/tips/reports/How-to-Use-a-Learning-Rate-Scheduler-in-Keras--VmlldzoyMjU2MTI3
- Tuning the Hyperparameters and Layers of Neural Network Deep Learning: https://www.analyticsvidhya.com/blog/2021/05/tuning-the-hyperparameters-and-layers-of-neural-network-deep-learning/
- GRIDSEARCHCV: https://machinelearningmastery.com/grid-search-hyperparameters-deep-learning-models-python-keras/
- https://studymachinelearning.com/keras-imagedatagenerator-with-flow_from_directory/

References:
- https://medium.com/free-code-camp/everything-you-need-to-know-to-master-convolutional-neural-networks-ef98ca3c7655
- https://medium.com/mlearning-ai/computer-vision-exploring-your-image-datasets-the-right-way-538c5ae8ca5d

CNN EXPLAINED
- https://www.youtube.com/watch?v=pj9-rr1wDhM

CNN
- https://machinelearningmastery.com/how-to-configure-image-data-augmentation-when-training-deep-learning-neural-networks/

DATA AUGMENTATION
- https://www.analyticsvidhya.com/blog/2020/08/image-augmentation-on-the-fly-using-keras-imagedatagenerator/#:~:text=Keras%20ImageDataGenerator%20is%20a%20gem,up%20on%20the%20overhead%20memory!
- https://www.analyticsvidhya.com/blog/2021/06/offline-data-augmentation-for-multiple-images/
- https://github.com/AISangam/Image-Augmentation-Using-OpenCV-and-Python/blob/master/Image%20Augmentaion%20Part1.py
- https://towardsdatascience.com/complete-image-augmentation-in-opencv-31a6b02694f5

HISTOGRAMS FOR COLORS IN IMAGE
- https://pyimagesearch.com/2021/04/28/opencv-image-histograms-cv2-calchist/
- https://web.stanford.edu/class/cs101/image-1-introduction.html#:~:text=Each%20of%20the%20red%2C%20green,in%20a%20shade%20of%20orange.

MANIFOLD
- https://towardsdatascience.com/manifolds-in-data-science-a-brief-overview-2e9dde9437e5

DROPOUT
- https://stackoverflow.com/questions/47892505/dropout-rate-guidance-for-hidden-layers-in-a-convolution-neural-network

FINE-TUNING
- https://towardsdatascience.com/fine-tuning-a-cnn-model-for-image-classification-a886e4b539b3
- https://www.analyticsvidhya.com/blog/2021/05/tuning-the-hyperparameters-and-layers-of-neural-network-deep-learning/
- https://machinelearningmastery.com/grid-search-hyperparameters-deep-learning-models-python-keras/

CROSS-VALIDATION KERAS
- https://towardsdatascience.com/increase-the-accuracy-of-your-cnn-by-following-these-5-tips-i-learned-from-the-kaggle-community-27227ad39554

GRIDSEARCH
- https://www.youtube.com/watch?v=TvB_3jVIHhg

TO TRY:
1. Get all the lists of loss and accuracy
2. get the mean for each values
3. plot them!!!

TO CHECK FOR MORE IDEAS FOR THE ARCHITECTURE
- https://machinelearningmastery.com/review-of-architectural-innovations-for-convolutional-neural-networks-for-image-classification/
- https://levelup.gitconnected.com/simple-image-classification-with-cnn-dd5ee3b725
- https://www.learndatasci.com/tutorials/convolutional-neural-networks-image-classification/
- https://www.guru99.com/convnet-tensorflow-image-classification.html

References:
- https://github.com/TetsumichiUmada/food_app
  - https://github.com/TetsumichiUmada/food_app/blob/master/1_food_app.ipynb
  - https://github.com/Pyligent/food101-image-classification/blob/master/food-101_EDA_Model_RestNet_final.ipynb
  - https://github.com/TetsumichiUmada/food_app/blob/master/0_dataset.ipynb

- https://github.com/AhmedMaghawry/Food-Image-Recognition/blob/master/Food_Detection.ipynb
- FoodApp: http://blog.stratospark.com/creating-a-deep-learning-ios-app-with-keras-and-tensorflow.html
- https://medium.com/@ksusorokina/image-classification-with-convolutional-neural-networks-496815db12a8
- http://blog.stratospark.com/creating-a-deep-learning-ios-app-with-keras-and-tensorflow.html
- Rejection class: 
  - https://hal.archives-ouvertes.fr/hal-02963115/document
  - https://ai.stackexchange.com/questions/24914/image-classification-need-method-to-classify-unknown-objects-as-trash-3d

- CNN:
  - https://medium.com/free-code-camp/everything-you-need-to-know-to-master-convolutional-neural-networks-ef98ca3c7655

-None of the above class approach:
  - https://arxiv.org/pdf/1511.03719.pdf

- Cat/Dog dataset:
  - https://www.kaggle.com/competitions/dogs-vs-cats/data?select=train.zip
  - https://data-flair.training/blogs/cats-dogs-classification-deep-learning-project-beginners/

- Data augmentation:
  - https://medium.com/mlearning-ai/albumentations-a-python-library-for-advanced-image-augmentation-strategies-752bff3a3da0

Check the following notebook done during class:
- _FDL_E04.ipynb_

Optimizers:
- https://towardsai.net/p/l/impact-of-optimizers-in-image-classifiers
- https://opt-ml.org/papers/2021/paper53.pdf
- https://keras.io/api/optimizers/
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8321140/
- https://www.lightly.ai/post/which-optimizer-should-i-use-for-my-machine-learning-project
- https://medium.com/geekculture/a-2021-guide-to-improving-cnns-optimizers-adam-vs-sgd-495848ac6008

# Refactoring the code:

* USE TYPE ANNOTATION:
  - https://towardsdatascience.com/type-annotations-in-python-d90990b172dc
  - https://blog.logrocket.com/understanding-type-annotation-python/

*   CONVERT REUSABLE CODE IN FUNCTIONS AND MOVE THEM TO THE UTILS.PY FILE

* YOU SHOULD ALSO SAVE THE FINAL DATASET WITH ALL THE PERFOMANCE MEASURES AND TIME SPENT FOR EACH MODEL:
  - https://github.com/alessandro-maccario/Industry-Lab-project-private/blob/main/Notebooks/Part2_Training.ipynb

**Comment by alessandro-maccario** [comment]
- [ ] Addizionare PCA nel transfer learning, unicamente come metodo di comparazione in termini di riduzione dei parametri. Non sarà il modello che poi utilizzeremo, ma solo come comparazione!

- [ ] Salva tutti i modelli e passa a migliorare il notebook di evaluation. 

- [ ] Dopodiché riprendi l'ultimo notebook di Industry Lab come riferimento per il notebook di testing e deployment!
- [ ] Inserisci valore di  cut off sulla percentuale se inferiore al 50% allora la risposta fornita dalla rete deve essere relativa alla classe di rigetto. 
- [ ] Per il deployment usa i seguenti riferimenti:
> https://www.freecodecamp.org/news/how-to-deploy-your-machine-learning-model-as-a-web-app-using-gradio/
> https://www.section.io/engineering-education/deploying-machine-learning-model-as-an-app-in-python-using-gradio/
> https://python.plainenglish.io/build-and-deploy-machine-learning-apps-with-gradio-the-easiest-way-to-create-a-web-user-interface-6f5219025da3
> https://medium.com/@kalyaniavhale7/tutorial-on-gradio-library-ecb8055923a1
> https://abidlabs.medium.com/deploy-a-machine-learning-model-in-seconds-56c400ecc0b3
> https://trojrobert.github.io/a-guide-for-deploying-and-serving-machine-learning-with-model-streamlit-vs-gradio/
> https://abdulsamodazeez.com/how-to-build-a-machine-learning-web-app-in-python-using-gradio
> https://www.youtube.com/watch?v=53x330upumY
